projects/mpg123/decode_fuzzer.cc

```diff
@@ -0,0 +1,57 @@
+#include <cstddef>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <vector>
+
+#include "mpg123.h"
+#include "byte_stream.h"
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+  static bool initialized = false;
+  if (!initialized) {
+    mpg123_init();
+    initialized = true;
+  }
+  int ret;
+  mpg123_handle* handle = mpg123_new(nullptr, &ret);
+  if (handle == nullptr) {
+    return 0;
+  }
+
+  ret = mpg123_open_feed(handle);
+  if (ret != MPG123_OK) {
+    mpg123_delete(handle);
+    return 0;
+  }
+
+  std::vector<uint8_t> output_buffer(mpg123_outblock(handle));
+
+  size_t output_written = 0;
+  // Initially, start by feeding the decoder more data.
+  int decode_ret = MPG123_NEED_MORE;
+  ByteStream stream(data, size);
+  while ((decode_ret != MPG123_ERR)) {
+    if (decode_ret == MPG123_NEED_MORE) {
+      if (stream.capacity() == 0) {
+        break;
+      }
+      const size_t next_size = std::min(stream.GetNextSizeT(), stream.capacity());
+      const uint8_t* next_input = stream.UncheckedConsume(next_size);
```

Comment:
I think the new version might have a similar problem with overflows (proably worse)
If  `data` 10 bytes long, and the first `next_input` is 4 bytes, when we pass it to `mpg123_decode`, `mpg123_decode` can actually read all 10 bytes without ASAN noticing. I would copy `next_input` into a new malloced array (ASAN knows that `x=malloc(1); x[1]` is invalid).

Reply:
Good point, I've never thought about setting up the allocations to make them ASAN friendly, this is a great idea.

PTAL